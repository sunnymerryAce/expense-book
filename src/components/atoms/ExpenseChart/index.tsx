import React from "react";
import * as chartjs from "chart.js";
import { HorizontalBar, ChartData } from "react-chartjs-2";
import convertHexToRgba from "@/util/functions/convertHexToRgba";

export interface Category {
  label: string;
  amount: number;
  budget: number;
  color: string;
}

export interface ExpenseChartProps {
  categories: Array<Category>;
}

const getBackgroundColors = (categories: Array<Category>): Array<string> => {
  return categories.map((category) => {
    const rgba = convertHexToRgba(category.color, 0.2);
    return rgba || "";
  });
};

const getBorderColors = (categories: Array<Category>): Array<string> => {
  return categories.map((category) => {
    const rgba = convertHexToRgba(category.color, 1);
    return rgba || "";
  });
};

const options: chartjs.ChartOptions = {
  responsive: true,
  legend: {
    display: false,
  },
  scales: {
    xAxes: [
      {
        ticks: {
          beginAtZero: true,
        },
      },
    ],
  },
};

const ExpenseChart: React.FC<ExpenseChartProps> = (props) => {
  const data: ChartData<chartjs.ChartData> = {
    labels: props.categories.map((category) => {
      return category.label;
    }),
    datasets: [
      // 出費
      {
        label: "Expense",
        data: props.categories.map((category) => {
          return category.amount;
        }),
        backgroundColor: getBackgroundColors(props.categories),
        borderColor: getBorderColors(props.categories),
        barPercentage: 1,
        borderWidth: 1,
      },
      // 予算
      {
        label: "Budget",
        data: props.categories.map((category) => {
          return category.budget;
        }),
        backgroundColor: props.categories.map((category) => {
          return category.amount <= category.budget
            ? "rgba(0, 0, 0, 0.1)"
            : "rgba(255, 0, 0, 0.5)";
        }),
        barPercentage: 0.5,
      },
    ],
  };

  return <HorizontalBar data={data} width={100} options={options} />;
};

export default ExpenseChart;
