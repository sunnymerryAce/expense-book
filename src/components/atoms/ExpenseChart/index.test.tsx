import React from "react";
import ExpenseChart from "./";
import { shallow } from "enzyme";

describe("<ExpenseChart />", () => {
  let container: any = null;
  const props = {
    chartItems: [
      {
        categoryName: "Food",
        amount: 24000,
        color: "#7CB342",
        budget: 30000,
      },
      {
        categoryName: "Cafe",
        amount: 7000,
        color: "#D81B60",
        budget: 5000,
      },
      {
        categoryName: "test",
        amount: 12000,
        color: "#FDD835",
        budget: 12000,
      },
    ],
  };

  afterEach(() => {
    container.unmount();
    container = null;
  });

  test("should match the snapshot", () => {
    container = shallow(<ExpenseChart {...props} />);
    expect(container.html()).toMatchSnapshot();
  });
});
