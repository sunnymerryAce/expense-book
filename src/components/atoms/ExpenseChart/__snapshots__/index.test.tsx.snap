// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpenseChart /> should match the snapshot 1`] = `"<canvas height=\\"150\\" width=\\"100\\"></canvas>"`;
